rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,439 @@
+
+**summary: string**
+
+A brief description of the plugin can then be displayed in the automatically generated API documentation.
+
+# Future features
+
+## Indexing stats
+
+Can we index statistics about our API as part of this system? For example, I'm dumping information about which api declarations are missing comments in the console.
+
+## Longer term approach to "plugin service folders"
+
+Using sub folders is a short term plan. A long term plan hasn't been established yet, but it should fit in with our folder structure hierarchy goals, along with
+any support we have for sharing services among a related set of plugins, that are not exposed as part of the public API.
+# Recommendations for writing comments
+
+## @link comments for the referenced type
+
+Core has a pattern of writing comments like this:
+
+```ts
+  /** {@link IUiSettingsClient} */
+  uiSettings: IUiSettingsClient;
+```
+
+I don't see the value in this. In the IDE, I can click on the IUiSettingsClient type and get directed there, and in the API doc system, the
+type will already be clickable. This ends up with a weird looking API:
+
+![image](../images/repeat_type_links.png)
+
+The plan is to make @link comments work like links, which means this is unneccessary information.
+
+I propose we avoid this kind of pattern.
+
+## Export every referenced type
+
+The docs system handles broken link warnings but to avoid breaking the ci, I suggest we turn this off initially. This will mean however,
+that we may miss situations where we are referencing a type that is not actually exported. This will cause a broken link in the docs
+system
+
+For example if your index.ts file has:
+```ts
+export type foo: string | AnInterface;
+```
+
+and does not also export `AnInterface`, this will be a broken link in the docs system.
+
+Until we have better ci tools to catch these mistakes, developers will need to export every referenced type.
+
+## Avoid `Pick` pattern
+
+Connected to the above, if you use `Pick`, there are two problems. One is that it's difficult for a developer to see the functionality
+available to them at a glance, since they would have to keep flipping from the interface definition to the properties that have been picked.
+
+The second potential problem is that you will have to export the referenced type, and in some situations, it's an internal type that isn't exported.
+
+![image](../images/api_doc_pick.png)
+
+# Open questions
+
+## Required attribute
+
+`isRequired` is an optional parameter that can be used to display a badge next to the API.
+We can mark function parameters that do not use `?` or `| undefined` as required. Open questions:
+
+1. Are we okay with a badge showing for `required` rather than `optional` when marking a parameter as optional is extra work for a developer, and `required` is the default?
+
+2. Should we only mark function parameters as `required` or interface/class parameters? Essentially, should any declaration that is not nullable
```

Comment:
IMO, we should treat "required" as the default and "optional" as the exception. It feels a lot more common for parameters, etc. to be required rather than optional.

Reply:
++ (cc @snide @glitteringkatie )